lib/report_portal/cucumber/formatter.rb

```diff
@@ -9,41 +9,66 @@ class Formatter
       def initialize(config)
         ENV['REPORT_PORTAL_USED'] = 'true'
 
-        @queue = Queue.new
-        @thread = Thread.new do
-          @report = ReportPortal::Cucumber::Report.new
-          loop do
-            method_arr = @queue.pop
-            @report.public_send(*method_arr)
-          end
-        end
-        @thread.abort_on_exception = true
+        setup_message_processing
 
         @io = config.out_stream
 
-        [:test_case_started, :test_case_finished, :test_step_started, :test_step_finished].each do |event_name|
+        [:test_case_started, :test_case_finished, :test_step_started, :test_step_finished, :test_run_finished].each do |event_name|
           config.on_event event_name do |event|
-            @queue.push([event_name, event, ReportPortal.now])
+            process_message(event_name, event)
           end
         end
-        config.on_event :test_run_finished, &method(:on_test_run_finished)
+        config.on_event(:test_run_finished) { finish_message_processing }
       end
 
       def puts(message)
-        @queue.push([:puts, message, ReportPortal.now])
+        process_message(:puts, message)
         @io.puts(message)
         @io.flush
       end
 
       def embed(*args)
-        @queue.push([:embed, *args, ReportPortal.now])
+        process_message(:embed, *args)
+      end
+
+      private
+
+      def report
+        @report ||= ReportPortal::Cucumber::Report.new
+      end
+
+      def setup_message_processing
+        return if use_same_thread_for_reporting?
+
+        @queue = Queue.new
+        @thread = Thread.new do
+          loop do
+            method_arr = @queue.pop
+            report.public_send(*method_arr)
+          end
+        end
+        @thread.abort_on_exception = true
       end
 
-      def on_test_run_finished(_event)
-        @queue.push([:done, ReportPortal.now])
```

Comment:
what method is finishing report now?

Reply:
test_run_finished

Reply:
I forgot to fix it in this PR and fixed in https://github.com/reportportal/agent-ruby/pull/44/